### PR TITLE
Use proper ref count handling when adding to x509 store

### PR DESCRIPTION
**Pull request by ColtonWilley**
# Description

Use proper ref count handling when adding to x509 store

Fixes github issue #8222

### CODE REVIEW

File: src/x509_str.c [hunk 1]
```diff
@@ -1408,24 +1408,30 @@ int wolfSSL_X509_STORE_add_cert(WOLFSSL_X509_STORE* store, WOLFSSL_X509* x509)
...
-                else {
-                    result = WOLFSSL_FATAL_ERROR;
+                result = wolfSSL_X509_up_ref(x509);
+                if (result == WOLFSSL_SUCCESS) {
+                    result = wolfSSL_sk_X509_push(store->trusted, x509);
+                    if (result > 0) {
+                        result = WOLFSSL_SUCCESS;
+                    }
+                    else {
+                        result = WOLFSSL_FATAL_ERROR;
```
Comment (SparkiDev) [inline src/x509_str.c:1418]: Should we be freeing x509 on failure to push?

Comment (ColtonWilley) [reply src/x509_str.c:1418]: Possibly, there are still additional issues with the implementation in this PR the automated tests exposed. I am currently working on resolving these new memory leaks.

File: src/x509_str.c [hunk 2]
```diff
@@ -1196,20 +1194,18 @@ void wolfSSL_X509_STORE_free(WOLFSSL_X509_STORE* store)
             }
 #if defined(OPENSSL_EXTRA)
             if (store->certs != NULL) {
-                wolfSSL_sk_X509_free(store->certs);
+                wolfSSL_sk_X509_pop_free(store->certs, wolfSSL_X509_free);
```
Comment (julek-wolfssl) [inline src/x509_str.c:1197]: Second param can be NULL. We should be recognizing the correct function to use internally.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: src/x509_str.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/x509_str.c:1413]: Very good catch @SparkiDev. We should call free to decrement the ref counter.

Comment (julek-wolfssl) [reply src/x509_str.c:1413]: In both cases.

Comment (ColtonWilley) [reply src/x509_str.c:1413]: Agreed, a great catch. The changes have been made.

**Comment by ColtonWilley** [comment]
Retest this please.

**Review by SparkiDev — APPROVED** [review]

**Review by douzzer — APPROVED** [review]
this all LGTM.

A separate issue, which you probably noticed yourself, is that we have a `wolfSSL_X509_up_ref()` but no corresponding `wolfSSL_X509_down_ref()`.  Hence, the `*free()` routines need to use _ad hoc_ `wolfSSL_RefDec(&store->ref, ...)` calls.

Cleanup opportunity for another time.

**Pull request merged**